receiver/k8seventsreceiver/receiver.go

```diff
@@ -0,0 +1,146 @@
+	ctx             context.Context
+	cancel          context.CancelFunc
+	obsrecv         *obsreport.Receiver
+}
+
+// newReceiver creates the Kubernetes events receiver with the given configuration.
+func newReceiver(
+	set component.ReceiverCreateSettings,
+	config *Config,
+	consumer consumer.Logs,
+	client k8s.Interface,
+) (component.LogsReceiver, error) {
+	transport := "http"
+
+	return &k8seventsReceiver{
+		settings:     set,
+		config:       config,
+		client:       client,
+		logsConsumer: consumer,
+		obsrecv: obsreport.NewReceiver(obsreport.ReceiverSettings{
+			ReceiverID:             config.ID(),
+			Transport:              transport,
+			ReceiverCreateSettings: set,
+		}),
+	}, nil
+}
+
+func (kr *k8seventsReceiver) Start(ctx context.Context, host component.Host) error {
+	kr.ctx, kr.cancel = context.WithCancel(ctx)
+
+	kr.settings.Logger.Info("Starting to watch namespaces for the events.")
+	if len(kr.config.Namespaces) == 0 {
+		kr.startWatch(corev1.NamespaceAll)
+	} else {
+		for _, ns := range kr.config.Namespaces {
+			kr.startWatch(ns)
+		}
+	}
+
+	return nil
+}
+
+func (kr *k8seventsReceiver) Shutdown(context.Context) error {
+	// Stop watching all the namespaces by closing all the stopper channels.
+	for _, stopperChan := range kr.stopperChanList {
+		close(stopperChan)
+	}
+	kr.cancel()
+	return nil
+}
+
+// Add the 'New Event' handler and trigger the watch for a specific namespace.
+func (kr *k8seventsReceiver) startWatch(ns string) {
+	stopperChan := make(chan struct{})
+	kr.stopperChanList = append(kr.stopperChanList, stopperChan)
+	kr.startWatchingNamespace(kr.client, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			ev := obj.(*corev1.Event)
+			kr.handleNewEvent(ev)
+		},
+	}, ns, stopperChan)
+}
+
+func (kr *k8seventsReceiver) handleNewEvent(ev *corev1.Event) {
+	if kr.allowEvent(ev) {
+		ld := k8sEventToLogData(kr.settings.Logger, ev)
+
+		ctx := kr.obsrecv.StartLogsOp(kr.ctx)
+		consumerErr := kr.logsConsumer.ConsumeLogs(ctx, ld)
+		kr.obsrecv.EndLogsOp(ctx, typeStr, 1, consumerErr)
+	}
+}
+
+// startWatchingNamespace creates an informer and starts
+// watching a specific namespace for the events.
+func (kr *k8seventsReceiver) startWatchingNamespace(
+	clientset k8s.Interface,
+	handlers cache.ResourceEventHandlerFuncs,
+	ns string,
+	stopper chan struct{},
+) {
+	client := clientset.CoreV1().RESTClient()
+	watchList := cache.NewListWatchFromClient(client, "events", ns, fields.Everything())
+	_, controller := cache.NewInformer(watchList, &corev1.Event{}, 0, handlers)
+	go controller.Run(stopper)
+}
+
+// Allow events with eventTimestamp(EventTime/LastTimestamp/FirstTimestamp)
+// not older than the current time by 'InitialLookback' so that
+// event flood can be avoided upon startup.
+func (kr *k8seventsReceiver) allowEvent(ev *corev1.Event) bool {
+	eventTimestamp := getEventTimestamp(ev)
+	return !eventTimestamp.Before(time.Now().Add(-kr.config.InitialLookback))
```

Comment:
I still don't understand why do we need `config.InitialLookback`. Can it be `!eventTimestamp.Before(kr.startTime)` if we want to avoid flood upon startup?

Reply:
>  I still don't understand why do we need `config.InitialLookback`. 

We have added this config option to allow users to get events having timestamps older than the receiver's start time.

>  Can it be `!eventTimestamp.Before(kr.startTime)`? Why do we use `time.Now()` if we want to avoid flood upon **startup** as stated above?

`!eventTimestamp.Before(kr.startTime)` can avoid flood upon startup but if some events with timestamp later than the receiver's start time are received again then we can't avoid that but as per the current logic `!eventTimestamp.Before(time.Now().Add(-kr.config.InitialLookback))`, we can avoid events having timestamp later than receiver's start time but older than the current time frame `time.Now().Add(-kr.config.InitialLookback)`.

Reply:
> but if some events with timestamp later than the receiver's start time are received again

Sorry I don't understand this. What kind of events can be received **again** that we have to drop? Can you please elaborate and provide an example?

Reply:
Let's consider the following scenario, 
where `InitialLookback` is 2 units and receiver starts at time `t`.
At the start time, events between `[t-2, t]` will be allowed. Now, at time `t+2` if any events having timestamp between `[t-2, t]` arrives then we are discarding it.
I am not sure of the use-case where this scenario will happen but this is just a programmatical precausion to discard any events having timestamp between `[t-2, t]` at time `t+2`.

If this doesn't seem necessary then I can use `kr.startTime` instead of `time.now()`.

Reply:
Given that `t = kr.startTime` in your example, I believe we can drop any events having timestamp older than `t`. I still don't understand what precaution you are trying to build with this logic.

Reply:
In the last comment, I was trying to describe the need for `InitialLookback` like if a user wants to allow older events having a timestamp between `[t-2, t]` then they can get those events by providing 2 units in the `InitialLookback`.
For e.x. If the collector stops and starts after some time then `InitialLookback` will allow the users to collect the events from that time interval.

And about `t = kr.startTime`, I don't have a use-case where duplicate events will arrive when the receiver is up and running. So, I have used `kr.startTime` instead of `time.now()`.

Reply:
> For e.x. If the collector stops and starts after some time then InitialLookback will allow the users to collect the events from that time interval.

Ok but if collector restarts within 5 seconds, and `InitialLookback` is 60s by default, we will be sending duplicate events happened within 55s before the restart. I would rather allow losing events during the restart than sending duplicates by default. Removing `InitialLookback` also simplifies the config interface which is always good. We can add it later if there is real need for that.

@jpkrohling WDYT?

Reply:
Agree -- keep it simple first, enhance in the future if needed.

Reply:
Okay, I have removed the `InitialLookback` config option.